provider/dns_domain_resource.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	cmv1 "github.com/openshift-online/ocm-sdk-go/clustersmgmt/v1"
+	"github.com/terraform-redhat/terraform-provider-rhcs/internal/ocm/resource"
+)
+
+type DNSDomainResourceType struct {
+}
+
+type DNSDomainResource struct {
+	collection         *cmv1.ClustersClient
+	dns_domains_client *cmv1.DNSDomainsClient
+}
+
+func (t *DNSDomainResourceType) NewResource(ctx context.Context, p tfsdk.Provider) (result tfsdk.Resource, diags diag.Diagnostics) {
+	// Cast the provider interface to the specific implementation:
+	// use it directly when needed.
+	parent := p.(*Provider)
+
+	// Get the collection of clusters
+	collection := parent.connection.ClustersMgmt().V1().Clusters()
+
+	//Create DNS Clients - not related to a cluster.
+	dns_domains_client := parent.connection.ClustersMgmt().V1().DNSDomains()
+
+	// Create the resource
+	result = &DNSDomainResource{
+		collection:         collection,
+		dns_domains_client: dns_domains_client,
+	}
+
+	return
+}
+
+func (t *DNSDomainResourceType) GetSchema(ctx context.Context) (result tfsdk.Schema,
+	diags diag.Diagnostics) {
+
+	result = tfsdk.Schema{
+		Description: "Identity provider.",
+		Attributes: map[string]tfsdk.Attribute{
+			"id": {
+				Description: "Unique identifier of the identity provider.",
+				Type:        types.StringType,
+				Computed:    true,
+			},
+		},
+	}
+
+	return
+}
+
+func (r *DNSDomainResource) Read(ctx context.Context,
+	request tfsdk.ReadResourceRequest, response *tfsdk.ReadResourceResponse) {
+
+	// Get the current state
+	state := &DNSDomainState{}
+	diags := request.State.Get(ctx, state)
+	response.Diagnostics.Append(diags...)
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	// Find the identity provider
+	dnsDomain := resource.NewDNSDomain(r.dns_domains_client)
+	get, err := dnsDomain.Get(state.ID.Value)
+	if err != nil && get.Status() == http.StatusNotFound {
+		tflog.Warn(ctx, fmt.Sprintf("DNS domain (%s) not found, removing from state",
+			state.ID.Value,
+		))
+		response.State.RemoveResource(ctx)
+		return
+	} else if err != nil {
+		response.Diagnostics.AddError(
+			fmt.Sprintf(
+				"Can't find DNS domain with identifier '%s'",
+				state.ID.Value,
+			),
+			err.Error(),
+		)
+		return
+	}
+
+	object := get.Body()
+
+	if id, ok := object.GetID(); ok {
+		state.ID = types.String{
+			Value: id,
+		}
+	}
+
+	// Save the state
+	diags = response.State.Set(ctx, state)
+	response.Diagnostics.Append(diags...)
+}
+
+func (r *DNSDomainResource) Create(ctx context.Context,
+	request tfsdk.CreateResourceRequest, response *tfsdk.CreateResourceResponse) {
+
+	// Get the plan
+	plan := &DNSDomainState{}
+	diags := request.Plan.Get(ctx, plan)
+	response.Diagnostics.Append(diags...)
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	dnsDomain := resource.NewDNSDomain(r.dns_domains_client)
```

Comment:
why `dns_domains_client` isn't internal created inside `resource/dns_domain`?

Reply:
We have discussed this. I do think we should have all clients created in side `resource/*` BUT we are not passing all the parameters needed down the chain to create the client elsewhere. We create clients via the parent (root provider), This is the only place where we have the information.